### PR TITLE
Dangling controllers cleanup at kubeturbo start

### PR DESCRIPTION
**Pull request by irfanurrehman**
Fixes https://vmturbo.atlassian.net/browse/OM-66590

~~This is on top of https://github.com/turbonomic/kubeturbo/pull/522. Only the last 2 commits are reviewable here.~~
#522 is merged.

### CODE REVIEW

**Comment by irfanurrehman** [comment]
Rebased!

File: pkg/discovery/worker/garbage_collector.go [hunk 1]
```diff
@@ -3,62 +3,108 @@ package worker
...
-	// Also cleanup immediately at startup
-	g.cleanupLeakedClonePods()
-	g.cleanupLeakedWrongSchedulerPods()
+func (g *GarbageCollector) StartCleanup() error {
+	glog.V(4).Info("Start garbage cleanup.")
+
+	// We cleanup the controllers before cleaning up the pods to ensure
+	// that the pending (invalid controller) pods are deleted at the right time.
+	if err := g.CleanupControllers(); err != nil {
+		return nil
```
Comment (ading1977) [inline pkg/discovery/worker/garbage_collector.go:72]: Why return `nil` when there is an error?

File: pkg/discovery/worker/garbage_collector.go [hunk 2]
```diff
@@ -110,3 +156,78 @@ func (g *GarbageCollector) isLeakedPod(pod api.Pod) bool {
 
 	return false
 }
+
+// The controllers cleanup is supposed to happen only at the startup
+// We fail in case of errors to force kubeturbo to restart and try the cleanup again
+func (g *GarbageCollector) CleanupControllers() error {
```
Comment (ading1977) [inline pkg/discovery/worker/garbage_collector.go:162]: Suggest renaming to something like `revertControllers` or `resetControllers`, or some other name to indicate that we are reverting back to the original state.

File: pkg/discovery/worker/garbage_collector.go [hunk 3]
```diff
@@ -110,3 +156,78 @@ func (g *GarbageCollector) isLeakedPod(pod api.Pod) bool {
...
+	if err := g.cleanupParents(); err != nil {
```
Comment (ading1977) [inline pkg/discovery/worker/garbage_collector.go:163]: Suggest renaming to something like `revertScheduler`

File: pkg/discovery/worker/garbage_collector.go [hunk 4]
```diff
@@ -110,3 +156,78 @@ func (g *GarbageCollector) isLeakedPod(pod api.Pod) bool {
...
+		return err
+	}
+
+	if err := g.cleanupGrandParents(); err != nil {
```
Comment (ading1977) [inline pkg/discovery/worker/garbage_collector.go:167]: Suggest renaming to something like `unpauseRollout`

File: pkg/action/executor/move_util.go [hunk 5]
```diff
@@ -478,20 +484,23 @@ func resourceRollout(client dynamic.ResourceInterface, obj *unstructured.Unstruc
 	return nil
 }
 
-// changeScheduler sets the scheduler value to default or unsets it to an invalid one
-func changeScheduler(client dynamic.ResourceInterface, obj *unstructured.Unstructured, valid bool) error {
+// ChangeScheduler sets the scheduler value to default or unsets it to an invalid one
+// This additionally adds or removes the garbage collection label on the resource
+func ChangeScheduler(client dynamic.ResourceInterface, obj *unstructured.Unstructured, valid bool) error {
```
Comment (enlinxu) [inline pkg/action/executor/move_util.go:489]: Can we also split this method to two methods, i.e., ChangeScheduler and RevertScheduler ? It's clearer to me compared to using valid / invalid (nothing is invalid here)

File: pkg/action/executor/move_util.go [hunk 6]
```diff
@@ -451,8 +451,9 @@ func getPodOwnersInfo(clusterScraper *cluster.ClusterScraper, pod *api.Pod,
 	return parent, nil, nsParentClient, nil, "", nil
 }
 
-// resourceRollout pauses/unpauses the rollout of a deployment or a deploymentconfig
-func resourceRollout(client dynamic.ResourceInterface, obj *unstructured.Unstructured, pause bool) error {
+// ResourceRollout pauses/unpauses the rollout of a deployment or a deploymentconfig
+// This additionally adds or removes the garbage collection label on the resource
+func ResourceRollout(client dynamic.ResourceInterface, obj *unstructured.Unstructured, pause bool) error {
```
Comment (enlinxu) [inline pkg/action/executor/move_util.go:456]: Similarly, let's split it into pauseRollout and resumeRollout. Using a boolean here maybe more error prone.

Comment (irfanurrehman) [reply pkg/action/executor/move_util.go:456]: I have a different opinion:
Its a lot of duplication of code and what we get in terms of the clarity is the `Pause` in the name of the function call in place of the name of argument. For example this is the implementation of `PauseRollout` (Which will be exact same implementation for `UnpauseRollout` except the flag value `true` in `PauseRollout`).
```
func PauseRollout(client dynamic.ResourceInterface, obj *unstructured.Unstructured) error {
	kind := obj.GetKind()
	name := obj.GetName()
	// This takes care of conflicting updates for example by operator
	// Ref: https://github.com/kubernetes/client-go/blob/master/examples/dynamic-create-update-delete-deployment/main.go
	retryErr := retry.RetryOnConflict(retry.DefaultRetry, func() error {
		objCopy, err := client.Get(context.TODO(), name, metav1.GetOptions{})
		if err != nil {
			return err
		}

		if err := unstructured.SetNestedField(objCopy.Object, true, "spec", "paused"); err != nil {
			return err
		}
		addGCLabel(objCopy)

		_, err = client.Update(context.TODO(), objCopy, metav1.UpdateOptions{})
		if err != nil {
			return err
		}
		return nil
	})

	if retryErr != nil {
		return fmt.Errorf("error setting 'spec.paused' to true for %s %s: %v", kind, name, retryErr)
	}
	return nil
}
```

To ensure clarity in such cases, I always try to use self documenting code  as below:
```
			pause := true
			unpause := false
```
and then use these names while passing the arguments.

File: pkg/action/executor/move_util.go (hunk 5 above)
Comment (irfanurrehman) [reply pkg/action/executor/move_util.go:489]: Please see the comment below (for which I did create an example). The same applies here.
Would a better naming (invalid -> incorrect/bad/somethingelse) explain it more? 
Meanwhile the intention is indeed set a name of a scheduler to something `invalid` or `bad`. That means if the cluster for some advanced reasons also has a scheduler named `turbo-scheduler`, we will need to use some other `invalid` name in place. `ChangeScheduler` and `RevertScheduler` does not make this explicit, rather might make it sound that we are using another scheduler to the reader of the code without context.

File: pkg/action/executor/move_util.go [hunk 7]
```diff
@@ -511,6 +520,29 @@ func changeScheduler(client dynamic.ResourceInterface, obj *unstructured.Unstruc
...
+	labels := obj.GetLabels()
+	if labels == nil {
+		// nothing to do
+		return
+	}
+	if _, exists := labels[TurboGCLabelKey]; exists {
+		delete(labels, TurboGCLabelKey)
+	}
+
+	obj.SetLabels(labels)
```
[the remark this answers is not in the document]
Comment (irfanurrehman) [reply pkg/action/executor/move_util.go]: Done!

File: pkg/action/executor/move_util.go [hunk 8]
```diff
@@ -478,20 +484,23 @@ func resourceRollout(client dynamic.ResourceInterface, obj *unstructured.Unstruc
...
+func ChangeScheduler(client dynamic.ResourceInterface, obj *unstructured.Unstructured, valid bool) error {
 	kind := obj.GetKind()
 	name := obj.GetName()
 	// This takes care of conflicting updates for example by operator
 	// Ref: https://github.com/kubernetes/client-go/blob/master/examples/dynamic-create-update-delete-deployment/main.go
 	retryErr := retry.RetryOnConflict(retry.DefaultRetry, func() error {
 		objCopy, err := client.Get(context.TODO(), name, metav1.GetOptions{})
 		if valid {
-			if err := unstructured.SetNestedField(objCopy.Object, "default-scheduler",
+			removeGCLabel(objCopy)
```
[the remark this answers is not in the document]
Comment (irfanurrehman) [reply pkg/action/executor/move_util.go]: Done!

File: pkg/action/executor/move_util.go [hunk 9]
```diff
@@ -462,6 +463,11 @@ func resourceRollout(client dynamic.ResourceInterface, obj *unstructured.Unstruc
 		if err != nil {
 			return err
 		}
+		if pause {
+			addGCLabel(objCopy)
+		} else {
+			removeGCLabel(objCopy)
```
[the remark this answers is not in the document]
Comment (irfanurrehman) [reply pkg/action/executor/move_util.go:473]: Done!

File: pkg/discovery/worker/garbage_collector.go (hunk 1 above)
Comment (irfanurrehman) [reply pkg/discovery/worker/garbage_collector.go]: Thanks for catching this. Updated!

File: pkg/discovery/worker/garbage_collector.go (hunk 2 above)
Comment (irfanurrehman) [reply pkg/discovery/worker/garbage_collector.go]: Done!

File: pkg/discovery/worker/garbage_collector.go (hunk 3 above)
Comment (irfanurrehman) [reply pkg/discovery/worker/garbage_collector.go]: Done!

File: pkg/discovery/worker/garbage_collector.go (hunk 4 above)
Comment (irfanurrehman) [reply pkg/discovery/worker/garbage_collector.go]: Done!

**Comment by irfanurrehman** [comment]
@ading1977 @enlinxu Updated. PTAL!

File: pkg/discovery/worker/garbage_collector.go [hunk 10]
```diff
@@ -110,3 +156,78 @@ func (g *GarbageCollector) isLeakedPod(pod api.Pod) bool {
...
+
+	return nil
+}
+
+func (g *GarbageCollector) revertSchedulers() error {
+	for _, parentRes := range supportedParents {
+		objList, err := g.dynClient.Resource(parentRes).Namespace("").List(context.TODO(), gcListOpts)
+		if err != nil {
+			// we fail forcing kubeturbo to restart and try this again
+			glog.Errorf("Error garbage cleaning parent controller for %s: %v", parentRes.String(), err)
```
Comment (ading1977) [inline pkg/discovery/worker/garbage_collector.go:179]: nit: "garbage cleaning parent controller" -> "reverting scheduler"

File: pkg/discovery/worker/garbage_collector.go [hunk 11]
```diff
@@ -110,3 +156,78 @@ func (g *GarbageCollector) isLeakedPod(pod api.Pod) bool {
...
+			return nil
+		}
+		for _, item := range objList.Items {
+			valid := true
+			err := commonutil.RetryDuring(executor.DefaultRetryLess, 0,
+				executor.DefaultRetrySleepInterval, func() error {
+					return executor.ChangeScheduler(g.dynClient.Resource(parentRes).Namespace(item.GetNamespace()), &item, valid)
+				})
+			if err != nil {
+				glog.Errorf("Error garbage cleaning controller for %s %s/%s: %v", item.GetKind(), item.GetNamespace(), item.GetName(), err)
```
Comment (ading1977) [inline pkg/discovery/worker/garbage_collector.go:193]: nit: "garbage cleaning parent controller" -> "reverting scheduler"

File: pkg/discovery/worker/garbage_collector.go [hunk 12]
```diff
@@ -110,3 +156,78 @@ func (g *GarbageCollector) isLeakedPod(pod api.Pod) bool {
...
+func (g *GarbageCollector) unpauseRollouts() error {
+	for _, grandParentRes := range supportedGrandParents {
+		objList, err := g.dynClient.Resource(grandParentRes).Namespace("").List(context.TODO(), gcListOpts)
+		if err != nil {
+			// This error could be because of mising deploymentConfig type from the cluster, so we don't fail here
+			if apierrors.IsNotFound(err) && strings.Contains(err.Error(), "the server could not find the requested resource") {
+				glog.V(3).Infof("Resource not found enabled on cluster while garbage cleaning controllers %s: %v", grandParentRes.String(), err)
+				continue
+			}
+			glog.Errorf("Error garbage cleaning grand parent controller for %s: %v", grandParentRes.String(), err)
```
Comment (ading1977) [inline pkg/discovery/worker/garbage_collector.go:211]: nit: "garbage cleaning grand parent controller" -> "unpausing rollout"

File: pkg/discovery/worker/garbage_collector.go [hunk 13]
```diff
@@ -110,3 +156,78 @@ func (g *GarbageCollector) isLeakedPod(pod api.Pod) bool {
...
+			return nil
+		}
+		for _, item := range objList.Items {
+			unpause := false
+			err := commonutil.RetryDuring(executor.DefaultRetryLess, 0,
+				executor.DefaultRetrySleepInterval, func() error {
+					return executor.ResourceRollout(g.dynClient.Resource(grandParentRes).Namespace(item.GetNamespace()), &item, unpause)
+				})
+			if err != nil {
+				glog.Errorf("Error garbage cleaning controller for %s %s/%s: %v", item.GetKind(), item.GetNamespace(), item.GetName(), err)
```
Comment (ading1977) [inline pkg/discovery/worker/garbage_collector.go:225]: nit: "garbage cleaning grand parent controller" -> "unpausing rollout"

**Review by ading1977 — APPROVED** [review]

File: pkg/discovery/worker/garbage_collector.go (hunk 10 above)
Comment (irfanurrehman) [reply pkg/discovery/worker/garbage_collector.go]: Updated!

File: pkg/discovery/worker/garbage_collector.go (hunk 11 above)
Comment (irfanurrehman) [reply pkg/discovery/worker/garbage_collector.go]: Updated!

File: pkg/discovery/worker/garbage_collector.go (hunk 12 above)
Comment (irfanurrehman) [reply pkg/discovery/worker/garbage_collector.go]: Updated!

File: pkg/discovery/worker/garbage_collector.go (hunk 13 above)
Comment (irfanurrehman) [reply pkg/discovery/worker/garbage_collector.go]: Updated!

**Comment by irfanurrehman** [comment]
@ading1977 Updated the nits. PTAL!

**Pull request merged**